tests/unit/test_solvers/test_rk4_solver.py

```diff
@@ -0,0 +1,247 @@
+        solution = solver.solve(model, t_eval, inputs={"rate": 0.1})
+        np.testing.assert_allclose(solution.y[0], np.exp(-0.1 * solution.t), rtol=1e-06)
+
+    def test_model_solver_with_external(self):
+        # Create model
+        model = pybamm.BaseModel()
+        domain = ["negative electrode", "separator", "positive electrode"]
+        var1 = pybamm.Variable("var1", domain=domain)
+        var2 = pybamm.Variable("var2", domain=domain)
+        model.rhs = {var1: -var2}
+        model.initial_conditions = {var1: 1}
+        model.external_variables = [var2]
+        model.variables = {"var1": var1, "var2": var2}
+        # No need to set parameters; can use base discretisation (no spatial
+        # operators)
+
+        # create discretisation
+        mesh = get_mesh_for_testing()
+        spatial_methods = {"macroscale": pybamm.FiniteVolume()}
+        disc = pybamm.Discretisation(mesh, spatial_methods)
+        disc.process_model(model)
+        # Solve
+        solver = pybamm.RK4Solver(N=20)
+        t_eval = np.linspace(0, 10, 100)
+        solution = solver.solve(model, t_eval, external_variables={"var2": 0.5})
+        np.testing.assert_allclose(solution.y[0], 1 - 0.5 * solution.t, rtol=1e-06)
+
+    def test_model_solver_ode_events(self):
+        model = pybamm.BaseModel()
+        # model.convert_to_format = "python"
+        whole_cell = ["negative electrode", "separator", "positive electrode"]
+        var = pybamm.Variable("var", domain=whole_cell)
+        model.rhs = {var: 0.1 * var}
+        model.initial_conditions = {var: 1}
+        model.events = [
+            pybamm.Event("2 * var = 2.5", pybamm.min(2 * var - 2.5)),
+            pybamm.Event("var = 1.5", pybamm.min(var - 1.5)),
+        ]
+        disc = get_discretisation_for_testing()
+        disc.process_model(model)
+
+        # Solve
+        solver = pybamm.RK4Solver(N=20)
+        t_eval = np.linspace(0, 10, 100)
+        solution = solver.solve(model, t_eval)
+        np.testing.assert_allclose(solution.y[0], np.exp(0.1 * solution.t), rtol=1e-06)
+        np.testing.assert_array_less(solution.y[0], 1.5)
+        np.testing.assert_array_less(solution.y[0], 1.25)
+
+    def test_code_gen(self):
+        # Create model
+        model = pybamm.BaseModel()
+        var = pybamm.Variable("var")
+        model.rhs = {var: 0.1 * var}
+        model.initial_conditions = {var: 1}
+        # No need to set parameters; can use base discretisation (no spatial operators)
+
+        # create discretisation
+        disc = pybamm.Discretisation()
+        model_disc = disc.process_model(model, inplace=False)
+        # Solve
+        solver = pybamm.RK4Solver(N=20)
+        t_eval = np.linspace(0, 1, 100)
+        solution = solver.solve(model_disc, t_eval)
+        opts = dict(main=True, mex=True)
+        solution.F.generate("gen.c", opts)
+        # To test the generated code run the following in the terminal
+        # gcc gen.c -o gen
```

Comment:
if code generation is going to be a feature of this solver, how to do this should be documented as part of the class